kimchi/cairo/src/runner/bytecode.rs

```diff
@@ -0,0 +1,106 @@
+// TODO(querolita):
+// - be able to index memory with any type (u256 in particular) not only usize
+
+use crate::runner::word::CairoWord;
+
+/// This data structure stores the memory of the program
+pub struct CairoBytecode {
+    /// length of the public memory
+    publen: u64,
+    /// full memory vector, None if non initialized
+    pub vect: Vec<Option<CairoWord>>,
+}
+
+impl CairoBytecode {
+    /// Create a new memory structure from a vector of u64
+    pub fn new(input: Vec<i128>) -> CairoBytecode {
+        // Initialized with the public memory (compiled instructions only)
+        // starts intentionally with a zero word for ease of testing
+        let mut aux = vec![0];
+        aux.extend(input);
+        CairoBytecode {
+            publen: (aux.len() - 1) as u64,
+            vect: aux.into_iter().map(|i| Some(CairoWord::new(i))).collect(),
+        }
+    }
+
+    /// Get size of the public memory
+    pub fn public(&self) -> u64 {
+        self.publen
+    }
+
+    /// Get size of the full memory including dummy 0th entry
+    pub fn size(&self) -> u64 {
+        (self.vect.len()) as u64
+    }
+
+    /// Enlarges memory with enough additional None slots if necessary before writing or reading
+    fn enlarge(&mut self, index: u64) {
+        // if you want to access an index of the memory but its size is less or equal than this
+        if self.size() <= index {
```

Comment:
you probably want `<` and not `<=`

Reply:
Mmm I do think I need `<=` here for a "weird" reason, let me explain. At first my vectors (memory, basically) started with the 0 index, as usual when programming. But then I realised that the tests one could extract from the Cairo playground started with a memory address of 1. That means, that every time I was checking a program, I needed to manually "adapt" the memory content to be one less value for the addresses that contained stack addresses inside. And this is a pain when you do it more than twice. So I just gave up and inserted a dummy "zero" entry at the beginning of the memory vector, so that useful entries started on position 1 (thus the second) as in the playground. So for example: say I have a program with 3 instructions. Then they will go from position 0 (with the dummy), then 1, 2 and 3. So one would have a 4-elements sized memory with only 3 "real" cells. Now suppose you want to write in the 4th (fake) position, meaning the fifth. Then you need to enlarge it by one. And since `self.size() = 4` and `index = 4`, then I need to include the `<=`. But when I get a "normal" test-suite that won't use the playground outputs, then I will be able to remove the equal and have a normal behaviour.